Setup metrics collection

lampajr opened the pull request:
Changes:
- [x] Setup pidstat to monitor CPU usage for running services
- [x] Setup pmap to measure the RSS over the time
- [x] Setup perf-stat to gather performance counter statistics
- [x] Setup vmstat to report virtual memory stats
- [x] Setup mpstat to report processors stats
- [x] Update perform-fights benchmark
- [x] Fix opentelemetry communication

TODOs (in another PR?):
- [ ] Setup application profiler, either `async profiler` or `perf record` based on the app running (native or not)

lampajr committed:
Setup metrics collection

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr commented:
@johnaohara @stalep I think this is already a good starting point, the only thing I am currently missing is async profiler or perf record to get the stacktrace of the running application but given that it would require some additional effort I'd like to do that in different PR.

With this PR I also changed the perform-fights benchmark as well as the setup such that now the opentelemetry communication is working also in our internal lab (which was not before - there were some errors).

johnaohara commented on benchmarks/perform-fights/perform-fights.hf.yaml:
```diff
@@ -83,17 +83,25 @@ phases:
+  #               - httpRequest:
+  #                   GET: /api/fights/randomlocation
+  #                   headers:
+  #                     accept: application/json
```
I think we should remove these lines if we are going to reuse the `performFight` scenario

lampajr replied:
Done

johnaohara commented on monitoring/otel-collector-config.yaml:
```diff
@@ -7,7 +7,7 @@ receivers:
 exporters:
   otlp/jaeger:
-    endpoint: $OTEL_HOSTNAME$:4317
+    endpoint: localhost:4317
```
Why change this to `localhost`? do we not need to be able to configure the exporter for different hosts, or will this always be on the same host?

lampajr replied:
Currently Jaeger and the Otel collector are running in the same server, therefore I think we can keep `localhost` for now.

johnaohara commented on profiling.yaml:
```diff
@@ -0,0 +1,112 @@
+      - repeat-until: HF_BENCHMARK_TERMINATED
+        then:
+        - sleep: 1s
+        - sh: echo -n "$(date +"%I:%M:%S %p")  " >> /tmp/FIGHTS_rss.log
+        - sh:
+            # NOTE: with docker this requires starting the containers with the same host user, this is now sets by default with
+            # --user "$(id -u):$(id -g)"
+            command: pmap -x ${{HOST.FIGHTS_REST_PID}} ${{PMAP_PARAMS:}} 2>/dev/null | grep total >> /tmp/FIGHTS_rss.log
```
This will create 2 lines per entry, one for timestamp, then another for the measurement. can we store in local variables and write one line to file.

lampajr replied:
Technically they should be on one line already as the timestamp is echoed with `-n` option which should not create a new line at the end. But I think we can still save it in a variable

johnaohara commented on superheroes.yaml:
```diff
@@ -137,6 +139,7 @@ scripts:
         ${{ CONTAINER_RUNTIME_PARAMS }}
         --memory ${{HEROES_REST_MEMORY}}
         --cpus ${{HEROES_REST_CPU}}
+        --user "$(id -u):$(id -g)"
```
why do we need to map user:group in the container?

lampajr replied:
This is actually required when using `docker` in order to make use of `pmap`. Otherwise the /proc/PID/smaps file would be owned by root and pmap will not have enough permission to access it and I would avoid to use sudo cmds. Do you see any issue in forcing the user:group?

lampajr replied:
I made it configurable because for instance for `jvm` builds they are specifying the USER in the Dockerfile so we cannot use our own user to startup the container

johnaohara commented on superheroes.yaml:
```diff
@@ -150,6 +153,9 @@ scripts:
       - regex: (?<HOST.HEROES_REST_POD_ID>[a-f0-9]{64}$)
         else:
         - abort: failed to capture pod id for Heroes rest service
+      - sh: ${{CONTAINER_RUNTIME}} inspect --format='{{.State.Pid}}' ${{HOST.HEROES_REST_POD_ID}}
+        then:
+        - set-state: HOST.HEROES_REST_PID
```
If the container has a wrapper script, does this get the PID of the java process, or the wrapper script?
The semeru builds use a wrapper script to start the java process inside the container

lampajr replied:
Good question, I don't know.. I will check it

lampajr replied:
Ok, I just tried a simple example and `inspect --format='{{.State.Pid}}'` will get the main process PID, i.e., the wrapper in the case of semeru builds 🫤

lampajr replied:
With a dirty trick https://github.com/RedHatPerf/superheroes-automation/pull/9/commits/2220169fa148310faf4bfd7bf4c69450600ed2db I was able to retrieve the correct pid when using semeru builds - looks working but it is assuming the main command is a `java` command which should be always true for semeru builds

lampajr committed:
Extract correct pid when entrypoint is a wrapper

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr commented on modes/assets/semeru/start-quarkus.sh:
```diff
@@ -11,7 +11,7 @@ else
     $JAVA_HOME/bin/java -XX:CRaCCheckpointTo=cr -Dopenj9.internal.criu.unprivilegedMode=true ${JAVA_OPTS} ${JAVA_OPTS_APPEND} -jar ${JAVA_APP_JAR} 1>out 2>err </dev/null &
 
-    sleep 5
+    sleep 10
```
I had to increase this sleep to 10s because I noticed that sometimes it takes more than 5s to startup.. therefore the server could be not available when trying to hit the endpoint

lampajr committed:
Make user group mapping configurable

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr committed:
Remove resource limits on mongo

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr committed:
Update random fighters benchmark to max throughput

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr committed:
Add strace disabled by default

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr commented:
With the latest commits I improved the system profiling setup:
- [Remove resource limits on mongo](https://github.com/RedHatPerf/superheroes-automation/pull/9/commits/ac709f7015583eaeab6519fc6b213fd583ddd73c): Removed the resource limit on the mongo-db such that it is not the bottleneck
- [Update random fighters benchmark to max throughput](https://github.com/RedHatPerf/superheroes-automation/pull/9/commits/18d7937c09d839adc84f424f3630af029a1e8013): update the benchmark for random fighters to get the max throughput
- [Add strace disabled by default](https://github.com/RedHatPerf/superheroes-automation/pull/9/commits/c3af40d36f12378dac31cf9d894cf301221fc487): Added `strace` in case we need to run it as well

lampajr commented:
As agreed, I'll proceed to merge this one
